### ISSUE TITLE
Matching representations to a prior distribution is wrong?

### ISSUE THREAD

**Issue by kosta-jo** [DuaneNielsen/DeepInfomaxPytorch#16]
@DuaneNielsen I think that you have not implemented adversarial matching of the distributions. You calculate PRIOR loss [here](https://github.com/DuaneNielsen/DeepInfomaxPytorch/blob/2fa707545db93ebec10366f43b98933cd611ba55/train.py#L47) with the y that comes from [here](https://github.com/DuaneNielsen/DeepInfomaxPytorch/blob/2fa707545db93ebec10366f43b98933cd611ba55/train.py#L90), so when you do loss.backward() [here](https://github.com/DuaneNielsen/DeepInfomaxPytorch/blob/2fa707545db93ebec10366f43b98933cd611ba55/train.py#L96) the weights of the encoder will be altered so that the PRIOR loss is minimized - i.e. the PRIOR loss from the paper is maximized since you are doing negative of the original loss. So you are altering the weights of the encoder so that you maximize the PRIOR loss from the paper, rather than minimizing it. Have I understood this wrong?

**Comment by DuaneNielsen** [comment]
Hey, thanks for the precise question... It certainly got me thinking...

Here's the relevant section from the paper.

![image](https://user-images.githubusercontent.com/11070508/143656892-a17ae45d-1ad9-4855-a040-dd64f9c1fe30.png)

If I read the center of equation (7) when the loss is expressed in argmin/argmax KL divergence it reads to me as...

"Find the parameters of the encoder that minimize the KL divergence between a prior and the output of the encoder, as computed by a discriminator, where the discriminator parameters maximizes the estimate of the divergence."

So I think that means that you are correct, and the code is correct?

Also, in the interest of precision, what change would you make to the code to resolve the issue in your mind?  Might be easier to just resolve this question empirically :)

**Comment by kosta-jo** [comment]
@DuaneNielsen I think that what you should do is first calculate ``y`` like you did [here](https://github.com/DuaneNielsen/DeepInfomaxPytorch/blob/2fa707545db93ebec10366f43b98933cd611ba55/train.py#L90), then you calculate prior loss like you did [here](https://github.com/DuaneNielsen/DeepInfomaxPytorch/blob/2fa707545db93ebec10366f43b98933cd611ba55/train.py#L47) (only prior loss, not the whole loss), but you should call ``.detach()``  on the ``y`` in this calculation. And doing ``loss.backward()``  on the prior loss would change the weights of the discriminator so that the loss from the paper is maximized. If you do not detach ``y`` you will change the weights of the encoder so that loss from the paper is maximized, which would be wrong, since the encoder wants to minimize that loss. So the next step is to alter the weights of the encoder to minimize the loss from the paper and you do that by passing y that is NOT detached through the discriminator which will give you ``D(y)``  and you calculate ``new_loss=Ep[log(1-D(y))]``  and you do ``new_loss.backward()`` to minimize it.

**Comment by Minhvt34** [comment]
> 

Hi, as your suggestion, I really want to clarify that,
You mean, we should use like 
    term_a = torch.log(self.prior_d(prior)).mean()
    term_b = torch.log(1.0 - self.prior_d(y.detach())).mean()
    PRIOR = - (term_a + term_b)*self.gamma
    new_loss = torch.log(1.0 - self.prior_d(y)).mean() #          y without detach

    return LOCAL + GLOBAL + PRIOR , new_loss       # new_loss also be returned to .backward() during training.

Do I understand your points exactly?
